### ISSUE TITLE
c++ auto convert Dot to -> when dealing with std smart pointer types 

### ISSUE THREAD

**Issue by bodgergely** [microsoft/vscode-cpptools#6441]
**Type: LanguageService**

**Steps to reproduce**
1. Type '.' after a "smart pointer" (`std::unique_ptr`/`std::shared_ptr`) variable and we see the members of the 'smart pointer' type.

**Expected behavior**
1. Type '.' after a "smart pointer" (`std::unique_ptr`/`std::shared_ptr`) variable and **show the members of the wrapped type instead of the smart pointer type**. When the member is chosen, convert `.` to `->`. 

**Additional context**

**Comment by Colengms** [comment]
Hi @bodgergely .  The C/C++ Extension currently supports this behavior.  I see that this works as expected using a simple example.

![image](https://user-images.githubusercontent.com/49173979/98171625-e3c8f580-1ea4-11eb-8c6b-6b4172f39580.gif)

Does a simple example such as this work for you?  If so, are you seeing an issue in a particular scenario?

**Comment by bodgergely** [comment]
@Colengms First of all thanks for getting back so quickly!
You are right, the issue comes up only with unique_ptr/shared_ptr. Would be really nice to extend this functionality somehow to work with these smart pointers. Most of the time people want to actually dereference instead of accessing the member of unique_ptr/shared_ptr.
Not sure if this is doable in a clean way, but less and less people use naked pointers so this will become more and more useful.

**Comment by bobbrow** [comment]
We currently have the same behavior as Visual Studio for smart pointers.  Are you seeing the `->` members listed in the auto-complete list when you type a `.` (as suggested in your first comment)?  I was pretty sure we didn't do that yet.

@Colengms, @sean-mcmanus do you know if there is an issue open on Developer Community for this?  I was thinking this had been requested before.

**Comment by bodgergely** [comment]
@bobbrow Thanks for getting back.
Yes, when typing `.` I get the members of the smart pointer instead of the members of the wrapped pointer type. I understand that this is correct behaviour and in 1% of the cases you might actually want to access a member of the smart pointer type like (`.get()`). And yes, it works the same way in Visual Studio 2019 as well.

Of course defaulting `.` to show the members of the wrapped type means that you won't automatically get suggestions of the smart pointer type so I am not sure if this can be solved in a clean way.

**Comment by Colengms** [comment]
I've opened a feature request against VS here: https://developercommunity.visualstudio.com/idea/1246258/extend-dot-to-arrow-conversion-feature-to-smart-po.html
(We share common code for IntelliSense with VS)

**Comment by bodgergely** [comment]
@Colengms This is awesome! I highly appreciate the attention I got for this. Speaks volumes of how cool VS Code community is becoming.

**Comment by kingsimba** [comment]
With plugin VISUAL ASSIST(from https://www.wholetomato.com/), Visual Studio can convert . to -> automatically.
With 1% chance when I do need '.', I will use Ctrl+Z to undo the conversion.

**Comment by sean-mcmanus** [comment]
@kingsimba You mean it does that for smart pointer types?

**Comment by kingsimba** [comment]
> @kingsimba You mean it does that for smart pointer types?

Yes, even for smart pointer types. And there is a kind notice to tell you how to undo it.

![image](https://user-images.githubusercontent.com/8407634/116396802-5dd8a680-a858-11eb-9f6e-6e831f34950f.png)

And there is another difference. MSVS will convert . to -> before typing anything else, while VSCode do the conversion after typing the function/variable name.
